### PR TITLE
Add DAC Codec to espnet-codec

### PR DESCRIPTION
**Pull request by wyh2000**
## What?

This PR aims to add a new codec architecture DAC to ESPnet. 
DAC is based on SoundStream design. Also, it contains some specific design to accelerate and optimize the training process.

In this repo, we add:
1. Periodic activation function
2. Quantizer dropout
3. Multi-scale multi-period multi-resolution discriminator
4. Factorized codes

## See also
DAC official code: https://github.com/descriptinc/descript-audio-codec
DAC paper: https://arxiv.org/pdf/2306.06546

### CODE REVIEW

File: espnet2/gan_codec/shared/quantizer/residual_vq.py [hunk 1]
```diff
@@ -83,9 +89,15 @@ def forward(
...
+        quantized, codes, commit_loss, quantization_loss = self.vq(x, n_q=n_q)
         bw = torch.tensor(n_q * bw_per_q).to(x)
-        return quantized, codes, bw, torch.mean(commit_loss)
+        return (
+            quantized,
+            codes,
+            bw,
+            torch.mean(commit_loss),
+            torch.mean(quantization_loss),
+        )
```
Comment (ftshijt) [inline espnet2/gan_codec/shared/quantizer/residual_vq.py:100]: Could you please verify that the change won't introduce potential issues to the soundstream/encodec modules?

File: espnet2/train/preprocessor.py [hunk 2]
```diff
@@ -400,7 +400,7 @@ def _pad_speech(self, speech):
                 constant_values=((0, 0), (0, self.audio_pad_value)),
             )
 
-        logging.info(speech.shape)
+        # logging.info(speech.shape)
```
Comment (ftshijt) [inline espnet2/train/preprocessor.py:403]:
```suggestion
```
We can simply remove it~

File: espnet2/gan_codec/shared/discriminator/msmpmb_discriminator.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+            ]
+        )
+        self.conv_post = WNConv1d(1024, 1, 3, 1, padding=1, act=False)
+        self.sample_rate = sample_rate
+        self.rate = rate
+
+    def forward(self, x):
+        # x = AudioSignal(x, self.sample_rate)
+        # x.resample(self.sample_rate // self.rate)
+        # x = x.audio_data
```
Comment (ftshijt) [inline espnet2/gan_codec/shared/discriminator/msmpmb_discriminator.py:52]:
```suggestion
```

File: espnet2/gan_codec/dac/dac.py [hunk 4]
```diff
@@ -0,0 +1,747 @@
...
+        bw = self.target_bandwidths[random.randint(0, max_idx)]
+
+        # Forward quantizer
+        quantized, _, _, commit_loss, quantization_loss = self.quantizer(
+            encoder_out, self.frame_rate, bw
+        )
+
+        # quantization_loss = self.l1_quantization_loss(
+        #     encoder_out, quantized.detach()
+        # ) + self.l2_quantization_loss(encoder_out, quantized.detach())
```
Comment (ftshijt) [inline espnet2/gan_codec/dac/dac.py:604]:
```suggestion
```

File: egs2/libritts/codec1/run.sh [hunk 5]
```diff
@@ -20,10 +20,12 @@ train_set=train-clean-460
 valid_set=dev-clean
 test_sets="dev-clean test-clean"
 
-train_config=conf/train_soundstream.yaml
+train_config=conf/train_dac.yaml
 inference_config=conf/decode.yaml
 
 ./codec.sh \
+    --stage 1 \
+    --stop_stage 6 \
```
Comment (ftshijt) [inline egs2/libritts/codec1/run.sh:28]:
```suggestion
```

**Review by ftshijt — COMMENTED** [review]
Many thanks for your great contribution. I mainly have one concern on the compatibility issue with existing soundstream/encodec after we change the core_vq part

**Comment by ftshijt** [comment]
@wyh2000 Please let me know when the PR is ready for final review, thanks!

**Comment by wyh2000** [comment]
> @wyh2000 Please let me know when the PR is ready for final review, thanks!

It's ready to be final review now.

**Comment by ftshijt** [comment]
LGTM! Thanks for your contribution~
